Add Zstd library

johnelse opened the pull request:
This just wraps the `zstd` binary.

mseri commented on compression/compression.ml:
```diff
@@ -0,0 +1,89 @@
+
+  (** Helper function to prevent double-closes of file descriptors *)
+  let close to_close fd =
+    if List.mem fd !to_close then Unix.close fd;
```
Close could raise `EBADF` (that should not happen), `EINTR` (this may happen), `EIO` (this may happen). Should we protect this close in a `try ... with` block to make sure that all the cleanup is performed?

mseri commented on compression/compression.ml:
```diff
@@ -0,0 +1,89 @@
+
+    finally
+      (fun () ->
+         let args = if mode = Compress then [] else ["--decompress"] @ [ "--stdout"; "--force" ] in
```
Why `--force` by default?

lindig commented:
I'm not a fan of extending xen-api-libs-transitional as it has developed into a grab bag of unrelated code. Why do we need this and could there be another way of packaging this?

gaborigloi commented:
Although this PR factors out the duplication nicely between the new zstd module and the old gzip one.
But we should get rid of xen-api-libs-transitional indeed.

johnelse commented:
For context, I'm working on adding zstd compression support for VM import/export to xapi, and adding a wrapper to the `zstd` binary is the first step. This seemed to be the simplest way to add it without functionally modifying the existing `Gzip` library, or duplicating a lot of code. If there's a better place to put this I'm happy to look into it.

mseri commented:
As much as I'd like for `xen-api-libs-transitional` and `xapi-stdext` to disappear, when we started the cleanup I thought we realized that some of them at least are bound to stay and one of them was the `gzip` wrapper. I cannot find yet the discussion, but there was a comment of @edwintorok explaining some of the reasons why in a past PR.

Because of the above, I think it is reasonable to consider this PR as a preliminary step to extract a `compressor` library and `compressor-gzip` and `compressor-zcat` sub-libraries. Especially given that we could easily drop the stdext dependencies in these by brutally copying the two functions that we use.

edwintorok commented:
Should we call this `xapi-compression`? It depends on forkexec and thus it is not general purpose.
Also please use `dune` files instead of `jbuild`, we're trying to update everything to use dune (xs-opam is currently at dune 1.4.0).

Having a separate process for long running CPU intensive tasks (whether using the binary, or link to a C library) makes it clearer where time is spent, if this was all inside XAPI then all the performance issues would get blamed on it.

johnelse commented:
> Should we call this `xapi-compression`?

Sure, I'll make that change.

> Also please use `dune` files instead of `jbuild`, we're trying to update everything to use dune (xs-opam is currently at dune 1.4.0).

Makes sense, but the rest of the repo currently uses jbuild. If you don't mind a mix then I can do this.

johnelse commented:
I've renamed `compression` to `xapi-compression` and converted everything except `http-svr` to dune. I had a go at converting `http-svr` as well but I couldn't stop the tests from hanging.

lindig commented on Makefile:
```diff
@@ -4,26 +4,26 @@ OPAM_LIBDIR?=$(DESTDIR)$(shell opam config var lib)
 
 release:
-	jbuilder build @install
+	dune build @install
```
This needs `--profile=release`

johnelse replied:
Done.

lindig commented on xapi-compression.opam:
```diff
@@ -0,0 +1,16 @@
+opam-version: "1.2"
```
This should be in Opam 2 format.

johnelse replied:
There were a few opam 2-isms missing from the new .opam files - I've updated both files.

lindig commented:
Something is wrong with this commit. The build of xs-opam is broken:

https://travis-ci.org/xapi-project/xs-opam/builds/465895055

Compiling the repo does work. So maybe the problem is in xs-opam - I will check the opam files in there.

#=== ERROR while compiling gzip.master ========================================#
# context              2.0.1 | linux/x86_64 | ocaml-base-compiler.4.06.1 | file:///mnt
# path                 ~/.opam/4.06/.opam-switch/build/gzip.master
# command              ~/.opam/4.06/bin/jbuilder build -p gzip -j 127
# exit-code            1
# env-file             ~/.opam/log/gzip-638-0deeee.env
# output-file          ~/.opam/log/gzip-638-0deeee.out
### output ###
# Info: creating file dune-project with this contents: (lang dune 1.4)
# File "gzip/dune", line 4, characters 13-29:
# 4 |   (libraries xapi-compression)
#                  ^^^^^^^^^^^^^^^^
# Error: Library "xapi-compression" not found.
# Hint: try: dune external-lib-deps --missing -p gzip @install
```

mseri commented:
Have you updated gzip opam file to depend on xapi-compression?

mseri commented:
Also you need to add xapi-compression to the packages in xs (and if you want zstd in xs-extra).
I would have probably moved gzip and zstd inside xapi-compression even if this was a breaking change.

lindig commented:
We need to update opam files in xs-opam. We are merging a fix.
